File: BfsAndDfs/P127WordLadder/WordLadder.java
package BfsAndDfs.P127WordLadder;
// Given two words (beginWord and endWord), and a dictionary's word list, find the length of shortest transformation sequence from beginWord to endWord, such that:

// Only one letter can be changed at a time.
// Each transformed word must exist in the word list.
// Note:

// Return 0 if there is no such transformation sequence.
// All words have the same length.
// All words contain only lowercase alphabetic characters.
// You may assume no duplicates in the word list.
// You may assume beginWord and endWord are non-empty and are not the same.

// Example 1:

// Input:
// beginWord = "hit",
// endWord = "cog",
// wordList = ["hot","dot","dog","lot","log","cog"]

// Output: 5

import java.util.*;

public class WordLadder {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        if (beginWord.length() != endWord.length() && !wordList.contains(endWord)) {
            return 0;
        }
        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);
        Set<String> set = new HashSet<>(wordList);
        int level = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i=0; i<size; i++) {
                String word = queue.poll();
                if (word.equals(endWord)) {
                    return level+1;
                }
                for (int j=0; j<word.length(); j++) {
                    StringBuilder sb = new StringBuilder(word);
                    for (char c = 'a'; c <= 'z'; c++) {
                        sb.setCharAt(j, c);
                        String cur = sb.toString();
                        if (set.contains(cur)) {
                            queue.offer(cur);
                            set.remove(cur);
                        }
                    }
                }
            }
            level++;
        }
        return 0;
    }
}

File: BfsAndDfs/P127WordLadder/WordLadder.py
import unittest
from collections import deque
from typing import List
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        q = deque([])
        s = set(wordList)
        q.append(beginWord)
        level = 0
        while(q):
            size = len(q)
            for _ in range(size):
                curr = q.popleft()
                if curr == endWord:
                    return level + 1
                for i in range(0, len(curr)):
                    for j in 'abcdefghijklmnopqrstuvwxyz':
                        tmp = curr[:i] + j + curr[i+1:]
                        if tmp in s:
                            q.append(tmp)
                            s.remove(tmp)
            level = level + 1
        return 0
class UnitTest(unittest.TestCase):
    def test_1(self):
        beginWord = "hit"
        endWord = "cog"
        wordList = ["hot","dot","dog","lot","log","cog"]
        self.assertEqual(Solution().ladderLength(beginWord, endWord, wordList), 5)
        
        wordList = ["hot","dot","dog","lot","log"]
        self.assertEqual(Solution().ladderLength(beginWord, endWord, wordList), 0)

if __name__ == "__main__":
    unittest.main()